File: ocaml/01-10/10.ml
(* Run-length encoding of a list. (easy) *)

let collect xs = List.fold_left 
  (fun (c, cs, all) x -> if x = c then (c, x::cs, all) else (x, [x], cs::all)) 
  (List.hd xs, [], []) 
  xs

let pack = function 
  | []  -> []
  | [x] -> [[x]]
  | xs  -> let (_, ys, zs) = collect xs in List.rev (ys::zs)

let encode xs = List.map (fun ls -> (List.length ls, List.hd ls)) (pack xs)

let _ = assert (
  (encode ["a";"a";"a";"a";"b";"c";"c";"a";"a";"d";"e";"e";"e";"e"]) = 
  [(4, "a"); (1, "b"); (2, "c"); (2, "a"); (1, "d"); (4, "e")])

File: prolog/01-10/10.pl
/*
Use the result of problem 1.09 to implement the so-called run-length encoding data compression method. Consecutive duplicates of elements are encoded as terms [N,E] where N is the number of duplicates of the element E.

Example:
?- encode([a,a,a,a,b,c,c,a,a,d,e,e,e,e],X).
X = [[4,a],[1,b],[2,c],[2,a],[1,d],[4,e]]
*/

/* 09.pl */
:- include('09').

/* 10.pl */

enc([], Enc, Enc).
enc([H|T], [H, N], Enc) :- N1 is N + 1, enc(T, [H, N1], Enc).

enc([], []).
enc([H|T], Enc) :- enc(T, [H, 1], Enc).

encall([], Acc, Encs) :- reverse(Acc, Encs).
encall([H|T], Acc, Encs) :- enc(H, E), encall(T, [E|Acc], Encs).

encode(Xs, Encs) :- 
  pack(Xs, Ys),
  encall(Ys, [], Encs).
